backend/src/main/java/fk/prof/backend/model/assignment/WorkAssignmentSchedule.java

```diff
@@ -0,0 +1,211 @@
+            if(value.tooEarly) {
+              return null; //Since this is a priority queue, no point checking subsequent entries if current entry indicates its too early
+            } else {
+              this.entries.poll(); //dequeue the entry. no point in keeping the entry around whether fetch was done on right time or it was a scheduling miss
+              if(value.workAssignment != null) {
+                this.assignedRecorders.add(recorderIdentifier);
+                return value.workAssignment;
+              }
+            }
+          }
+        } catch (Exception ex) {
+          //TODO: increment some metric somewhere
+          logger.error("Unexpected error when getting fetching work assignment from schedule", ex);
+        } finally {
+          entriesLock.unlock();
+        }
+      } else {
+        //TODO: increment some metric somewhere
+        logger.warn("Timeout while acquiring lock for fetching work assignment from schedule");
+      }
+    } catch (InterruptedException ex) {
+      //TODO: increment some metric somewhere
+      logger.warn("Interrupted while acquiring lock for fetching work assignment from schedule");
+    }
+    return null;
+  }
+
+  public static class ScheduleEntry implements Comparable<ScheduleEntry> {
+    private final static Logger logger = LoggerFactory.getLogger(ScheduleEntry.class);
+
+    private final Recorder.WorkAssignment.Builder workAssignmentBuilder;
+    private final long offsetFromReferenceInNanos;
+
+    public ScheduleEntry(Recorder.WorkAssignment.Builder workAssignmentBuilder, long offsetFromReferenceInNanos) {
+      this.workAssignmentBuilder = Preconditions.checkNotNull(workAssignmentBuilder);
+      this.offsetFromReferenceInNanos = offsetFromReferenceInNanos;
+    }
+
+    /**
+     * @param elapsedTimeSinceSchedulingInNanos
+     * @param minAcceptableDelayInSecs
+     * @param maxAcceptableDelayInSecs
+     * @return value of schedule entry with appropriate flags set to indicate if its too early or too late to receive work assignment
+     */
+    ScheduleEntryValue getValue(long elapsedTimeSinceSchedulingInNanos, int minAcceptableDelayInSecs, int maxAcceptableDelayInSecs) {
+      int remainingDelayInSecs = (int)((offsetFromReferenceInNanos - elapsedTimeSinceSchedulingInNanos) / NANOS_IN_SEC);
+      if (remainingDelayInSecs < minAcceptableDelayInSecs) {
+        logger.error(String.format("Scheduling miss for work_id=%d, remaining delay=%d", workAssignmentBuilder.getWorkId(), remainingDelayInSecs));
+        return new ScheduleEntryValue(null, false, true);
+      }
+      if (remainingDelayInSecs > maxAcceptableDelayInSecs) {
+        logger.debug(String.format("Too early to hand over work assignment for work_id=%d, remaining delay=%d", workAssignmentBuilder.getWorkId(), remainingDelayInSecs));
+        return new ScheduleEntryValue(null, true, false);
+      }
+      return new ScheduleEntryValue(
+          workAssignmentBuilder
+              .clone()
```

Comment:
Necessary to clone because of how protobuf lib is implemeted in java. Builder is useless once build is called, so need to clone before calling build